olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp

```diff
@@ -156,18 +150,19 @@ client::CancellationToken VersionedLayerClientImpl::GetData(
 
     auto data_task =
         [=](client::CancellationContext context) mutable -> DataResponse {
+      boost::optional<int64_t> version;
       if (!request.GetDataHandle()) {
         auto version_response = GetVersion(request.GetBillingTag(),
                                            request.GetFetchOption(), context);
         if (!version_response.IsSuccessful()) {
           return version_response.GetError();
         }
-        request.WithVersion(version_response.GetResult().GetVersion());
+        version = version_response.GetResult().GetVersion();
       }
 
       return repository::DataRepository::GetVersionedData(
-          std::move(catalog), std::move(layer_id), std::move(request), context,
-          std::move(settings));
+          std::move(catalog), std::move(layer_id), version.get(),
+          std::move(request), context, std::move(settings));
```

Comment:
if request has no data handle, version is not initialized and we will try to call `version.get()`